src/Mod/Sketcher/Gui/DrawSketchController.h

```diff
@@ -239,7 +241,6 @@ class DrawSketchController
                 static_cast<OnViewParameterVisibility>(hGrp->GetInt("OnViewParameterVisibility"));
         }
 
-        OnViewParameterVisibility onViewParameterVisibility;

```

Comment:
Changing of access specifier should trigger some kind of alarm. It is rarely the right decision.

Here you could have used the visibility() member function to get the value, or made an:
```
        bool isVisibility(OnViewParameterVisibility visibility) const
        {
            return onViewParameterVisibility == visibility;
        }

```

The reason why this member is private, is to avoid that the visibility can be changed from outside this class. That is important because this class has the responsibility for managing the visibility. Allowing change from the outside is not proper encapsulation as it risks losing control over its responsibility.

You may argue that other (inner) classes, or even implementation only classes (classes defined within a source file, not in a header) have public data members. Yes this is true. Most often the reason for this is a mixture of laziness and perceived reduced associated risk. That is specially the case in implementation-only classes that often are POD alike.

However, when a member is already private, changing it to public for direct access has little justification, if any.

Reply:
They are not principles for the sake of principles. It is proper definition of interfaces, which show the intend on how they should be used. Perhaps you wouldn't try to set the visibility from outside (now it is a read only from the outside). But perhaps someone else, less familiar with the code would, ignoring that this object is responsible. That person may even have a different need and given that the member is public, why not change the visibility from outside the class? It is more convenient... less code to write... If people start working like that, at the end they will be continuously fixing the code, as others will introduce behaviour that they do not expect.

It is the encapsulation boundary that tells different developers that the member is not intended for modification from the outside, and which marks that the object is responsible. Developers wanting to modify the visibility from the outside, will hit a compilation error, which should wake them up. Seeing themselves trying to move the member to have public access specifier, should serve as a waking up call that what they are doing is wrong. It is also much better detectable for a maintainer reviewing a PR.

The case where there is a setter and a getter, which you claim is read/write anyway, is also not that black and white. The main problem with a public data member is that it defines interface. Changing the behaviour of the class, requires changing all accesses to the data member. If one writes a getter and a setter from the outset, changing behaviour is transparent for client code.

I am not leaving these kind of comments to annoy you. I am trying to let you know the reasons that motivate one direction or the other. Programming is not difficult. Doing it right is hard.

Perhaps another line of argumentation: You should consider why on earth language designers, almost in every object oriented language (I know) came with the need for accessors and access specifiers, when it would be so easy to have everything public. Hopefully, you won't believe they were chosen to design a language, or their language was adopted by millions, just because they were clueless about what they were doing or "out of principles". If you want to dedicate professionally to this, you should start questioning why the language has the features it has.

Reply:
> That person may even have a different need and given that the member is public, why not change the visibility from outside the class?

But again. Why would someone want to change this variable if he doesn't need to? If someone need to write this variable, he will create a getter and that's it.

> The main problem with a public data member is that it defines interface. Changing the behaviour of the class, requires changing all accesses to the data member. If one writes a getter and a setter from the outset, changing behaviour is transparent for client code.

Well this is true and so I see the point of setter/getter with this argument. Thanks for pointing it out.

Reply:
> But again. Why would someone want to change this variable if he doesn't need to? If someone need to write this variable, he will create a getter and that's it.

If the member is public and somebody thinks it is a good idea to do it, for example because he wants to change visibility after some event happening in the controller. This person has an incentive to want to write to it, this person is not going to create a getter. This person is going to write to it directly. The consequence is that the control logic stops being encapsulated in the class. The problem with losing encapsulation, is that the class starts losing control about when visibility is changed. So the automation mechanism embedded in it may fail. The decision to write to it is most likely a bad decision. But the interface of the class does not show intent. Without showing intent, the probability that another developer makes a mistake is much higher.

It generally goes like this: developer A created the class and incorporated the logic in the class but was lazy to write the getter/setter and put the data member public. Developer B needs to enable visualisation at some event. Instead of incorporating the logic into the class (for example, by creating a function that convey this persons intend and takes into account, within the class how the logic needs to be adapted to take into account this event), because the data member is public, writes to it. Reviewer C does not realise the mistake (writing to a public data member is very much normal). Visualisation works at times. Developer D in bug fixing mode may remove the code of developer B, or may even write to the variable at other code chunk outside the class to "fix" the bug (or make it not reproducible anymore, but the bug arising at other circumstances, as it is not fixed, the bug is controlling directly from outside without accounting for the logic already in).

All this does not even start if there is a getter. A developer should think twice before writing a setter for a private data member (at least should look very well into the class and the embedded logic). A reviewer is way more likely to realise the issue of a setter being added, or a variable changing access specifier, than when it is just writing to a public variable. 

The biggest metaphorical pan to hit the head of a developer and wake him wake up is a compiler error. That lack of direct accessibility that encapsulation provides is the metaphorical pan.

Reply:
Thanks for the example it makes things clearer.